### PR TITLE
make findnext work for utf8 strings (and other iterators)

### PR DESCRIPTION
**Pull request by SimonDanisch**
... and why not use closures on 0.6, so that there is actually only one real implementation of findnext.
A simple benchmark suggests that there are no speed regression:

``` Julia
using BenchmarkTools
a = rand(10000)
v = 42.0
a[end] == v
b1 = @benchmark simon.findnext($a, $v, $(1))
b2 = @benchmark master.findnext($a, $v, $(1))
judge(minimum(b1), minimum(b2))
-> invariant
```

### CODE REVIEW

File: base/array.jl [hunk 1]
```diff
@@ -970,12 +970,7 @@ julia> findnext(A,3)
 ```
 """
 function findnext(A, start::Integer)
```
Comment (nalimilan) [inline base/array.jl]: Use the short function syntax since this now fits in one line.

File: base/array.jl [hunk 2]
```diff
@@ -1061,13 +1051,13 @@ julia> findnext(isodd, A, 2)
 0
 ```
 """
-function findnext(testf::Function, A, start::Integer)
-    for i = start:length(A)
-        if testf(A[i])
-            return i
-        end
+function findnext(testf::Function, A, idx::Integer)
+    while !done(A, idx)
```
Comment (nalimilan) [inline base/array.jl]: Unfortunately, I don't think we guarantee that the `state` is equivalent to indices for all indexable iterators. Maybe we need to add this requirement to the "indexable" interface?

Comment (TotalVerb) [inline base/array.jl]: Would `state2ind` and `ind2state` functions fix this, like `sub2ind` and `ind2sub`?

File: base/array.jl (hunk 1 above)
Comment (SimonDanisch) [inline base/array.jl]: very well ;)

File: base/array.jl (hunk 2 above)
Comment (nalimilan) [inline base/array.jl]: Probably, but that would be yet another function to implement for custom types. I guess we could have it fall back to `identity` so that only iterators which deviate from the standard need to implement it.

**Comment by nalimilan** [comment]
Could you also tackle `findprev` at the same time? Doesn't make sense to have it behave differently.

**Comment by SimonDanisch** [comment]
Done!
I needed to move `prevind(::AbstractArray, ::Integer)` to `abstractarray.jl`, since it wasn't defined early enough for `findprev` to work.
Well, I guess for now it's okay that this only works with integer indices. In the long run, shouldn't this also work with the `iterable` interface?
Improving on that would mean to accept other index types, define `prevind` for them and I guess `zero(::IndexType)` ?. Not sure how we can type `start/idx` to do that without ambiguities.

File: base/array.jl [hunk 3]
```diff
@@ -1210,11 +1188,13 @@ julia> findprev(isodd, A, 3)
...
 ```
 """
-function findprev(testf::Function, A, start::Integer)
-    for i = start:-1:1
-        testf(A[i]) && return i
+function findprev(testf::Function, A, idx::Integer)
+    while idx > 0
+        elem, _ = next(A, idx)
+        testf(elem) && return idx
+        idx = prevind(A, idx)
```
Comment (stevengj) [inline base/array.jl]: This seems to assume that all iterables implement `prevind`, when you have only defined it for `AbstractArray` (and strings).

**Comment by stevengj** [comment]
`findnext` and `findprev` should also work on ordered indexables that aren't `AbstactArray` subtypes, e.g. tuples.

**Comment by SimonDanisch** [comment]
True! Is `prevind`/`nextind` the right function to use here? If yes, I guess the way forward would be to include a fall back like `prevind(A, idx::Integer) = Int(idx)-1`.

**Comment by SimonDanisch** [comment]
Done just that!

File: base/abstractarray.jl [hunk 4]
```diff
@@ -690,8 +690,8 @@ start(A::AbstractArray) = (@_inline_meta; itr = eachindex(A); (itr, start(itr)))
 next(A::AbstractArray,i) = (@_propagate_inbounds_meta; (idx, s) = next(i[1], i[2]); (A[idx], (i[1], s)))
 done(A::AbstractArray,i) = (@_propagate_inbounds_meta; done(i[1], i[2]))
 
-nextind(s::AbstractArray    , i::Integer) = Int(i)+1
-prevind(s::AbstractArray    , i::Integer) = Int(i)-1
+nextind(s, i::Integer) = Int(i)+1
```
Comment (stevengj) [inline base/abstractarray.jl]: Why the conversion to `Int`?  It makes `i = nextind(s, i)` potentially non-type-stable.

File: base/array.jl (hunk 2 above)
Comment (nalimilan) [inline base/array.jl]: @stevengj What you do think of this problem?

File: base/abstractarray.jl (hunk 4 above)
Comment (SimonDanisch) [inline base/abstractarray.jl]: Good call... I just blindly copied that ;) I guess it stems from the fact, that a lot of interfaces don't accept anything else but `Int`.
I wish it wouldn't be like that, especially for GPU codegen.

**Comment by SimonDanisch** [comment]
I changed the `prevind`/`nextind` definition in `abstractarrays.jl`. The `idx::Integer`-> `Int(idx)` combo seems widespread in the strings code. Changing all this in the scope of this PR doesn't seem appropriate ;)

**Comment by nalimilan** [comment]
I think there's still an issue regarding the conflation of state and indices.

**Comment by SimonDanisch** [comment]
I don't think its in the scope of this PR to rewrite/modify the `Iterator` interface to make this implementation clean.
Can we agree that this version works strictly better than the old version and move the iterator problem into a new issue?

File: base/array.jl [hunk 5]
```diff
@@ -969,14 +969,7 @@ julia> findnext(A,3)
...
 """
-function findnext(A, start::Integer)
-    for i = start:length(A)
-        if A[i] != 0
-            return i
-        end
-    end
-    return 0
-end
+findnext(A, start::Integer) = findnext(x-> x!=0, A, start)
```
Comment (nalimilan) [inline base/array.jl]: Weird spacing before `->` and in other similar code below.

File: base/array.jl [hunk 6]
```diff
@@ -1061,13 +1048,13 @@ julia> findnext(isodd, A, 2)
...
-        if testf(A[i])
-            return i
-        end
+function findnext(testf::Function, A, idx::Integer)
+    while !done(A, idx)
+        lastidx = idx
+        elem, idx = next(A, idx)
+        testf(elem) && return lastidx
     end
-    return 0
```
Comment (nalimilan) [inline base/array.jl]: Unrelated change?

Comment (nalimilan) [inline base/array.jl]: More importantly: you assume that `idx` is an `Int`, or the function will be type unstable. Need `zero(idx)`.

File: base/array.jl [hunk 7]
```diff
@@ -1160,12 +1142,8 @@ julia> findprev(A, 1, 1)
...
 ```
 """
-function findprev(A, v, start::Integer)
-    for i = start:-1:1
-        A[i] == v && return i
-    end
-    return 0
-end
+findprev(A, v, start::Integer) = findprev(x-> x==v, A, start)
+
```
Comment (nalimilan) [inline base/array.jl]: Why a new line break?

File: test/arrayops.jl [hunk 8]
```diff
@@ -386,8 +387,26 @@ a = [0,1,2,3,0,1,2,3]
...
+@test findprev(isodd, (2,4,5,3,9,2,0), 7) == 5
 @test findprev(isodd, [2,4,5,3,9,2,0], 7) == 5
 @test findprev(isodd, [2,4,5,3,9,2,0], 2) == 0
+@test findprev(isodd, (2,4,5,3,9,2,0), 2) == 0
+
+# Utf8 string
+str = "𝐴 ∈ ℝⁿˣⁿ, 𝐯 ∈ ℝⁿ, λᵢ ∈ ℝ: 𝐯"
+idx = findnext(str, 'λ', 1)
+@test str[idx] == 'λ'
+idx = findnext(str,'ᵢ', idx)
```
Comment (nalimilan) [inline test/arrayops.jl]: Inconsistent spacing here and below.

**Comment by nalimilan** [comment]
> Can we agree that this version works strictly better than the old version and move the iterator problem into a new issue?

I don't think we can really merge it without a validation by core devs. Even if it's innocuous, this PR introduces the idea that states and indices are the same thing. I have no opposition to that, but it needs to be made explicit.

**Comment by TotalVerb** [comment]
`Dict`?

**Comment by SimonDanisch** [comment]
What does "making it explicit" mean in this context? Open an issue? Add TODOs?

So  I think there are 3 issues:
1) What to return when nothing is found
2) how to type the arguments
3) how to reverse

1) could be solved in a type stable way with `Nullable`, which would be a breaking change and should be consistently applied for all kind of find/match functions.
Another option would be defining `zero(idx::IDXType)`... This doesn't seem like a good idea at all, though.
2) I guess without traits we don't really have any other option as to leave the idx arg untyped.
3) extend the iterator interface with something like `prev`, which only needs to be implement for non integer states and errors for non reversible iterators.

File: base/array.jl (hunk 5 above)
Comment (tkelman) [inline base/array.jl]: spaces both before and after `->` would be more common

**Comment by tkelman** [comment]
Pretty sure there are enough types where state and index differ that this is not a correct definition to use in general.

File: base/array.jl (hunk 2 above)
Comment (nalimilan) [inline base/array.jl]: @StefanKarpinski Opinions? Cf. also https://github.com/JuliaLang/julia/issues/19332.

**Comment by nalimilan** [comment]
It's too bad the state vs. index issue has blocked this PR. Since the `findnext` family of functions is defined to work with linear indices, I think the way forward is to increase the input index `idx` by one at each iteration, and return that instead of the iteration state. Hopefully the compiler will be able to eliminate the redundant computation when `next` does exactly that. How does that sound?

**Comment by JeffBezanson** [comment]
Superseded by #23592.
